multiaccounts/accounts/database.go

```diff
@@ -602,57 +639,100 @@ func (db *Database) getAccountByAddress(tx *sql.Tx, address types.Address) (*Acc
 	return accounts[0], nil
 }
 
-func (db *Database) deleteKeypair(tx *sql.Tx, keyUID string) error {
-	keypairs, err := db.getKeypairs(tx, keyUID)
-	if err != nil && err != sql.ErrNoRows {
+func (db *Database) markAccountRemoved(tx *sql.Tx, address types.Address, clock uint64) error {
+	if tx == nil {
+		return errDbTransactionIsNil
+	}
+
+	_, err := db.getAccountByAddress(tx, address)
+	if err != nil {
 		return err
 	}
 
-	if len(keypairs) == 0 {
-		return ErrDbKeypairNotFound
+	query, err := tx.Prepare(`
+		UPDATE
+			keypairs_accounts
+		SET
+			removed = TRUE,
+			clock = ?
+		WHERE
+			address = ?
+	`)
+	if err != nil {
+		return err
+	}
+	defer query.Close()
+
+	_, err = query.Exec(clock, address)
+	return err
+}
+
+// Marking keypair as removed, will delete related keycards.
+func (db *Database) markKeypairRemoved(tx *sql.Tx, keyUID string, clock uint64) error {
+	if tx == nil {
+		return errDbTransactionIsNil
+	}
+
+	keypair, err := db.getKeypairByKeyUID(tx, keyUID)
+	if err != nil {
+		return err
+	}
+
+	for _, acc := range keypair.Accounts {
+		if acc.Removed {
+			continue
+		}
+		err = db.markAccountRemoved(tx, acc.Address, clock)
+		if err != nil {
+			return err
+		}
 	}
 
 	query := `
-		DELETE
-		FROM
+		UPDATE
 			keypairs
+		SET
+			removed = TRUE,
+			clock = ?
 		WHERE
 			key_uid = ?
 	`
 
-	if tx == nil {
-		_, err := db.db.Exec(query, keyUID)
+	stmt, err := tx.Prepare(query)
+	if err != nil {
 		return err
 	}
+	defer stmt.Close()
 
-	stmt, err := tx.Prepare(query)
+	_, err = stmt.Exec(clock, keyUID)
 	if err != nil {
 		return err
 	}
-	defer stmt.Close()
 
-	_, err = stmt.Exec(keyUID)
+	err = db.deleteAllKeycardsWithKeyUID(tx, keyUID)
 	return err
 }
 
-func (db *Database) GetKeypairs() ([]*Keypair, error) {
-	return db.getKeypairs(nil, "")
+func (db *Database) GetKeypairs(includeRemoved bool) ([]*Keypair, error) {
+	return db.getKeypairs(nil, "", includeRemoved)
 }
 
+// Returns keypair if it is not marked as removed and its accounts which are not marked as removed.
 func (db *Database) GetKeypairByKeyUID(keyUID string) (*Keypair, error) {
 	return db.getKeypairByKeyUID(nil, keyUID)
 }
 
-func (db *Database) GetAccounts() ([]*Account, error) {
-	return db.getAccounts(nil, types.Address{})
+func (db *Database) GetAccounts(includeRemoved bool) ([]*Account, error) {
```

Comment:
I was thinking about that and couldn't decide which approach to follow and then opted in for that one (with 51:49 :) ), where user decides whether to get removed accounts as well or not at the place of usage, instead having two functions like `getAccounts` and `getAllAccounts` or `getAccountsIncludingRemoved`. 

Anyhow I don't mind to change that, but let's hear from other reviewers what do they think about this?
(personally I would keep it as it's now, but let's have a majority)

Reply:
Sure! This is just an opinion, I normally also like to have a single API with parameters and "force" the user to know what data he is (or isn't) getting back, but in this particular case I think including removed accounts would be unexpected and undesirable for almost all users.

Reply:
I'm for a separate function, whereas the old one can be renamed to `getActiveAccounts` or something like this, and a new one. Another approach would be to have a separate call which would only fetches removed accounts and we deal with merging of both collections on the client side (if necessary). It might help with less confusing names though adds some work on the caller side.